File: sdks/node/src/plugins/claudecode-contract.ts
import os from 'node:os';
import path from 'node:path';
import { isObject, parseStrictJsonObject, type JsonObject } from './strict-json.js';

export const AGENT_KEY = 'claudecode';
export const CONFIG_FILE = 'settings.json';
export const GUARD_SCRIPT = 'guard.js';
export const AUDIT_SCRIPT = 'hook.js';
export const HOOK_TIMEOUT_SECONDS = 10;
export const GUARD_STATUS = 'Checking Elydora agent state';
export const AUDIT_STATUS = 'Recording Elydora tool use';

const HOOK_EVENTS = new Set([
  'SessionStart', 'Setup', 'InstructionsLoaded', 'UserPromptSubmit',
  'UserPromptExpansion', 'MessageDisplay', 'PreToolUse', 'PermissionRequest',
  'PostToolUse', 'PostToolUseFailure', 'PostToolBatch', 'PermissionDenied',
  'Notification', 'SubagentStart', 'SubagentStop', 'TaskCreated', 'TaskCompleted',
  'Stop', 'StopFailure', 'TeammateIdle', 'ConfigChange', 'CwdChanged',
  'FileChanged', 'WorktreeCreate', 'WorktreeRemove', 'PreCompact', 'PostCompact',
  'SessionEnd', 'Elicitation', 'ElicitationResult',
]);

const COMMON_HANDLER_KEYS = ['if', 'once', 'statusMessage', 'timeout'];
const HANDLER_KEYS: Readonly<Record<string, readonly string[]>> = {
  command: [
    ...COMMON_HANDLER_KEYS,
    'args',
    'async',
    'asyncRewake',
    'command',
    'rewakeMessage',
    'rewakeSummary',
    'shell',
    'type',
  ],
  prompt: [...COMMON_HANDLER_KEYS, 'continueOnBlock', 'model', 'prompt', 'type'],
  agent: [...COMMON_HANDLER_KEYS, 'model', 'prompt', 'type'],
  http: [...COMMON_HANDLER_KEYS, 'allowedEnvVars', 'headers', 'type', 'url'],
  mcp_tool: [...COMMON_HANDLER_KEYS, 'input', 'server', 'tool', 'type'],
};

export interface ClaudeHandler extends JsonObject {
  readonly type: 'command' | 'prompt' | 'agent' | 'http' | 'mcp_tool';
}

export interface ClaudeGroup extends JsonObject {
  readonly matcher?: string;
  readonly hooks: ClaudeHandler[];
}

export type ClaudeHooks = Record<string, ClaudeGroup[]>;

export interface ClaudeDocument {
  readonly exists: boolean;
  readonly filePath: string;
  readonly root: JsonObject;
  readonly hooks: ClaudeHooks;
  readonly hooksDisabled: boolean;
  readonly raw?: string;
}

export interface RenderedClaudeDocument {
  readonly document: ClaudeDocument;
  readonly changed: boolean;
  readonly next?: string;
}

export interface ClaudeRuntimeContract {
  readonly agentId: string;
  readonly guardPath: string;
  readonly auditPath: string;
}

interface RuntimeReference {
  readonly agentId: string;
  readonly scriptPath: string;
}

function samePath(left: string, right: string): boolean {
  const normalizedLeft = path.resolve(left);
  const normalizedRight = path.resolve(right);
  return process.platform === 'win32'
    ? normalizedLeft.toLowerCase() === normalizedRight.toLowerCase()
    : normalizedLeft === normalizedRight;
}

function sameAgentId(left: string, right: string): boolean {
  return process.platform === 'win32'
    ? left.toLowerCase() === right.toLowerCase()
    : left === right;
}

function requireKnownKeys(value: JsonObject, allowed: readonly string[], label: string): void {
  const keys = new Set(allowed);
  const extra = Object.keys(value).find((key) => !keys.has(key));
  if (extra) throw new Error(`${label} contains unsupported field "${extra}"`);
}

function requireNonEmptyString(value: unknown, field: string, label: string): string {
  if (typeof value !== 'string' || value.length === 0) {
    throw new Error(`${label} field "${field}" must be a non-empty string`);
  }
  return value;
}

function optionalString(value: JsonObject, field: string, label: string): void {
  if (value[field] !== undefined && typeof value[field] !== 'string') {
    throw new Error(`${label} field "${field}" must be a string`);
  }
}

function optionalNonEmptyString(value: JsonObject, field: string, label: string): void {
  if (value[field] !== undefined) requireNonEmptyString(value[field], field, label);
}

function optionalBoolean(value: JsonObject, field: string, label: string): void {
  if (value[field] !== undefined && typeof value[field] !== 'boolean') {
    throw new Error(`${label} field "${field}" must be a boolean`);
  }
}

function validateCommonHandlerFields(value: JsonObject, label: string): void {
  optionalString(value, 'if', label);
  optionalString(value, 'statusMessage', label);
  optionalBoolean(value, 'once', label);
  if (value.timeout !== undefined && (
    typeof value.timeout !== 'number'
    || !Number.isFinite(value.timeout)
    || value.timeout <= 0
  )) {
    throw new Error(`${label} timeout must be a positive finite number`);
  }
}

function validateStringArray(value: unknown, field: string, label: string): string[] {
  if (!Array.isArray(value) || value.some((item) => typeof item !== 'string')) {
    throw new Error(`${label} field "${field}" must be an array of strings`);
  }
  return value as string[];
}

function validateHandler(value: unknown, event: string, group: number, index: number): ClaudeHandler {
  const label = `Claude Code settings handler hooks.${event}[${group}].hooks[${index}]`;
  if (!isObject(value)) throw new Error(`${label} must be an object`);
  if (typeof value.type !== 'string' || !HANDLER_KEYS[value.type]) {
    throw new Error(`${label} has unsupported type "${String(value.type)}"`);
  }
  requireKnownKeys(value, HANDLER_KEYS[value.type], label);
  validateCommonHandlerFields(value, label);
  if (value.type === 'command') {
    requireNonEmptyString(value.command, 'command', label);
    if (value.args !== undefined) validateStringArray(value.args, 'args', label);
    optionalBoolean(value, 'async', label);
    optionalBoolean(value, 'asyncRewake', label);
    optionalNonEmptyString(value, 'rewakeMessage', label);
    optionalNonEmptyString(value, 'rewakeSummary', label);
    if (value.shell !== undefined && value.shell !== 'bash' && value.shell !== 'powershell') {
      throw new Error(`${label} field "shell" must be "bash" or "powershell"`);
    }
  } else if (value.type === 'prompt' || value.type === 'agent') {
    requireNonEmptyString(value.prompt, 'prompt', label);
    optionalString(value, 'model', label);
    if (value.type === 'prompt') optionalBoolean(value, 'continueOnBlock', label);
  } else if (value.type === 'http') {
    const rawUrl = requireNonEmptyString(value.url, 'url', label);
    let url: URL;
    try {
      url = new URL(rawUrl);
    } catch (error) {
      throw new Error(`${label} field "url" must be a valid URL`, {
        cause: error instanceof Error ? error : new Error(String(error)),
      });
    }
    if (!['http:', 'https:'].includes(url.protocol) || !url.hostname) {
      throw new Error(`${label} field "url" must use HTTP or HTTPS`);
    }
    if (value.headers !== undefined && (
      !isObject(value.headers)
      || Object.values(value.headers).some((item) => typeof item !== 'string')
    )) {
      throw new Error(`${label} field "headers" must map names to strings`);
    }
    if (value.allowedEnvVars !== undefined) {
      const allowedEnvVars = validateStringArray(value.allowedEnvVars, 'allowedEnvVars', label);
      if (allowedEnvVars.some((item) => item.length === 0)) {
        throw new Error(`${label} field "allowedEnvVars" must contain non-empty strings`);
      }
    }
  } else {
    requireNonEmptyString(value.server, 'server', label);
    requireNonEmptyString(value.tool, 'tool', label);
    if (value.input !== undefined && !isObject(value.input)) {
      throw new Error(`${label} field "input" must be an object`);
    }
  }
  return value as ClaudeHandler;
}

function validateGroup(value: unknown, event: string, index: number): ClaudeGroup {
  const label = `Claude Code settings matcher group hooks.${event}[${index}]`;
  if (!isObject(value)) throw new Error(`${label} must be an object`);
  requireKnownKeys(value, ['hooks', 'matcher'], label);
  if (value.matcher !== undefined) {
    if (typeof value.matcher !== 'string') throw new Error(`${label} matcher must be a string`);
  }
  if (!Array.isArray(value.hooks)) throw new Error(`${label} must contain a hooks array`);
  return {
    ...value,
    hooks: value.hooks.map((handler, handlerIndex) => (
      validateHandler(handler, event, index, handlerIndex)
    )),
  } as ClaudeGroup;
}

function readHooks(value: unknown): ClaudeHooks {
  if (value === undefined) return {};
  if (!isObject(value)) throw new Error('Claude Code settings field "hooks" must be an object');
  const hooks: ClaudeHooks = {};
  for (const [event, groups] of Object.entries(value)) {
    if (!HOOK_EVENTS.has(event)) {
      throw new Error(`Claude Code settings contains unsupported hook event "${event}"`);
    }
    if (!Array.isArray(groups)) {
      throw new Error(`Claude Code settings field "hooks.${event}" must be an array`);
    }
    hooks[event] = groups.map((group, index) => validateGroup(group, event, index));
  }
  return hooks;
}

export function parseClaudeDocument(filePath: string, raw: string): ClaudeDocument {
  const root = parseStrictJsonObject(raw, `Claude Code user settings at ${filePath}`);
  if (root.disableAllHooks !== undefined && typeof root.disableAllHooks !== 'boolean') {
    throw new Error('Claude Code settings field "disableAllHooks" must be a boolean');
  }
  return {
    exists: true,
    filePath,
    root,
    hooks: readHooks(root.hooks),
    hooksDisabled: root.disableAllHooks === true,
    raw,
  };
}

export function createClaudeDocument(filePath: string): ClaudeDocument {
  return { exists: false, filePath, root: {}, hooks: {}, hooksDisabled: false };
}

export function buildClaudeGroup(scriptPath: string, statusMessage: string): ClaudeGroup {
  return {
    hooks: [{
      type: 'command',
      command: process.execPath,
      args: [scriptPath],
      timeout: HOOK_TIMEOUT_SECONDS,
      statusMessage,
    }],
  };
}

function exactManagedGroup(group: ClaudeGroup): boolean {
  return Object.keys(group).length === 1 && Object.hasOwn(group, 'hooks');
}

function runtimeReference(scriptPath: string, scriptName: string): RuntimeReference | undefined {
  if (!path.isAbsolute(scriptPath) || path.basename(scriptPath) !== scriptName) return undefined;
  const agentDirectory = path.dirname(scriptPath);
  if (!samePath(path.dirname(agentDirectory), path.join(os.homedir(), '.elydora'))) return undefined;
  const agentId = path.basename(agentDirectory);
  return agentId && agentId !== '.' && agentId !== '..' ? { agentId, scriptPath } : undefined;
}

function managedReference(
  handler: ClaudeHandler,
  scriptName: string,
  statusMessage: string,
  includeLegacy = false,
): RuntimeReference | undefined {
  const keys = Object.keys(handler).sort().join('|');
  if (keys === 'args|command|statusMessage|timeout|type'
    && handler.type === 'command'
    && handler.timeout === HOOK_TIMEOUT_SECONDS
    && handler.statusMessage === statusMessage
    && typeof handler.command === 'string'
    && samePath(handler.command, process.execPath)
    && Array.isArray(handler.args)
    && handler.args.length === 1
    && typeof handler.args[0] === 'string') {
    return runtimeReference(handler.args[0], scriptName);
  }
  if (!includeLegacy || keys !== 'command|type' || handler.type !== 'command') return undefined;
  const match = typeof handler.command === 'string'
    ? /^node "([^"\r\n]+)"$/.exec(handler.command)
    : undefined;
  return match ? runtimeReference(match[1], scriptName) : undefined;
}

function removeFromGroups(
  groups: readonly ClaudeGroup[],
  scriptName: string,
  statusMessage: string,
  agentId?: string,
): { groups: ClaudeGroup[]; removed: boolean } {
  const result: ClaudeGroup[] = [];
  let removed = false;
  for (const group of groups) {
    const handlers = group.hooks.filter((handler) => {
      const reference = managedReference(handler, scriptName, statusMessage, true);
      const owned = reference !== undefined
        && (agentId === undefined || sameAgentId(reference.agentId, agentId));
      if (owned) removed = true;
      return !owned;
    });
    if (handlers.length > 0) result.push({ ...group, hooks: handlers });
    else if (!exactManagedGroup(group)) result.push({ ...group, hooks: [] });
  }
  return { groups: result, removed };
}

export function removeManagedClaudeHooks(hooks: ClaudeHooks, agentId?: string): ClaudeHooks {
  const next: ClaudeHooks = Object.fromEntries(
    Object.entries(hooks).map(([event, groups]) => [event, [...groups]]),
  );
  for (const [event, scriptName, statusMessage] of [
    ['PreToolUse', GUARD_SCRIPT, GUARD_STATUS],
    ['PostToolUse', AUDIT_SCRIPT, AUDIT_STATUS],
    ['PostToolUseFailure', AUDIT_SCRIPT, AUDIT_STATUS],
  ] as const) {
    if (!next[event]) continue;
    const result = removeFromGroups(next[event], scriptName, statusMessage, agentId);
    if (!result.removed) continue;
    if (result.groups.length > 0) next[event] = result.groups;
    else delete next[event];
  }
  return next;
}

function managedEvent(event: string): readonly [string, string] | undefined {
  if (event === 'PreToolUse') return [GUARD_SCRIPT, GUARD_STATUS];
  if (event === 'PostToolUse' || event === 'PostToolUseFailure') {
    return [AUDIT_SCRIPT, AUDIT_STATUS];
  }
  return undefined;
}

function entirelyManaged(document: ClaudeDocument): boolean {
  if (!document.exists || !Object.keys(document.root).every((key) => key === 'hooks')) return false;
  const events = Object.entries(document.hooks);
  if (events.length === 0) return false;
  let handlers = 0;
  for (const [event, groups] of events) {
    const contract = managedEvent(event);
    if (!contract || groups.length === 0) return false;
    for (const group of groups) {
      if (!exactManagedGroup(group)
        || group.hooks.length === 0
        || group.hooks.some((handler) => (
          !managedReference(handler, contract[0], contract[1], true)
        ))) return false;
      handlers += group.hooks.length;
    }
  }
  return handlers > 0;
}

export function renderClaudeDocument(
  document: ClaudeDocument,
  hooks: ClaudeHooks,
): RenderedClaudeDocument {
  if (!document.exists && Object.keys(hooks).length === 0) {
    return { document, changed: false };
  }
  if (Object.keys(hooks).length === 0 && entirelyManaged(document)) {
    return { document, changed: true };
  }
  const root: JsonObject = { ...document.root };
  if (Object.keys(hooks).length > 0) root.hooks = hooks;
  else delete root.hooks;
  const next = `${JSON.stringify(root, null, 2)}\n`;
  return { document, changed: next !== document.raw, next };
}

function referencesForEvent(
  groups: readonly ClaudeGroup[],
  scriptName: string,
  statusMessage: string,
): Map<string, RuntimeReference[]> {
  const references = new Map<string, RuntimeReference[]>();
  for (const group of groups) {
    if (!exactManagedGroup(group)) continue;
    for (const handler of group.hooks) {
      const reference = managedReference(handler, scriptName, statusMessage);
      if (!reference) continue;
      const key = process.platform === 'win32'
        ? reference.agentId.toLowerCase()
        : reference.agentId;
      references.set(key, [...(references.get(key) ?? []), reference]);
    }
  }
  return references;
}

export function claudeRuntimeContracts(hooks: ClaudeHooks): ClaudeRuntimeContract[] {
  const guards = referencesForEvent(hooks.PreToolUse ?? [], GUARD_SCRIPT, GUARD_STATUS);
  const successes = referencesForEvent(hooks.PostToolUse ?? [], AUDIT_SCRIPT, AUDIT_STATUS);
  const failures = referencesForEvent(
    hooks.PostToolUseFailure ?? [],
    AUDIT_SCRIPT,
    AUDIT_STATUS,
  );
  const contracts: ClaudeRuntimeContract[] = [];
  for (const [key, guard] of guards) {
    const success = successes.get(key);
    const failure = failures.get(key);
    if (guard.length !== 1 || success?.length !== 1 || failure?.length !== 1) continue;
    if (!samePath(success[0].scriptPath, failure[0].scriptPath)) continue;
    contracts.push({
      agentId: guard[0].agentId,
      guardPath: guard[0].scriptPath,
      auditPath: success[0].scriptPath,
    });
  }
  return contracts;
}

File: sdks/go/cmd/elydora/plugins/claudecode_contract.go
package plugins

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"net/url"
	"reflect"
	"sort"
)

const (
	claudeAgentKey           = "claudecode"
	claudeConfigFile         = "settings.json"
	claudeGuardScript        = "guard.js"
	claudeAuditScript        = "hook.js"
	claudeHookTimeout        = float64(10)
	claudeGuardStatusMessage = "Checking Elydora agent state"
	claudeAuditStatusMessage = "Recording Elydora tool use"
)

var claudeHookEvents = stringSet(
	"SessionStart", "Setup", "InstructionsLoaded", "UserPromptSubmit",
	"UserPromptExpansion", "MessageDisplay", "PreToolUse", "PermissionRequest",
	"PostToolUse", "PostToolUseFailure", "PostToolBatch", "PermissionDenied",
	"Notification", "SubagentStart", "SubagentStop", "TaskCreated",
	"TaskCompleted", "Stop", "StopFailure", "TeammateIdle", "ConfigChange",
	"CwdChanged", "FileChanged", "WorktreeCreate", "WorktreeRemove",
	"PreCompact", "PostCompact", "SessionEnd", "Elicitation",
	"ElicitationResult",
)

var claudeHandlerKeys = map[string]map[string]struct{}{
	"command": stringSet(
		"if", "once", "statusMessage", "timeout", "args", "async",
		"asyncRewake", "command", "rewakeMessage", "rewakeSummary",
		"shell", "type",
	),
	"prompt": stringSet(
		"if", "once", "statusMessage", "timeout", "continueOnBlock", "model",
		"prompt", "type",
	),
	"agent": stringSet(
		"if", "once", "statusMessage", "timeout", "model", "prompt", "type",
	),
	"http": stringSet(
		"if", "once", "statusMessage", "timeout", "allowedEnvVars", "headers",
		"type", "url",
	),
	"mcp_tool": stringSet(
		"if", "once", "statusMessage", "timeout", "input", "server", "tool",
		"type",
	),
}

type claudeGroup struct {
	object   map[string]any
	handlers []map[string]any
}

type claudeHooks map[string][]claudeGroup

type claudeDocument struct {
	exists        bool
	filePath      string
	root          map[string]any
	hooks         claudeHooks
	hooksDisabled bool
	raw           []byte
}

type claudeRenderedDocument struct {
	document *claudeDocument
	changed  bool
	next     []byte
	remove   bool
}

type claudeRuntimeContract struct {
	agentID   string
	guardPath string
	auditPath string
}

func cloneClaudeObject(value map[string]any) map[string]any {
	clone := make(map[string]any, len(value))
	for key, item := range value {
		clone[key] = item
	}
	return clone
}

func cloneClaudeHooks(source claudeHooks) claudeHooks {
	clone := make(claudeHooks, len(source))
	for event, groups := range source {
		clone[event] = append([]claudeGroup(nil), groups...)
	}
	return clone
}

func validateClaudeKnownKeys(
	value map[string]any,
	allowed map[string]struct{},
	label string,
) error {
	fields := make([]string, 0, len(value))
	for field := range value {
		fields = append(fields, field)
	}
	sort.Strings(fields)
	for _, field := range fields {
		if _, ok := allowed[field]; !ok {
			return fmt.Errorf(`%s contains unsupported field %q`, label, field)
		}
	}
	return nil
}

func requireClaudeString(value map[string]any, field, label string) (string, error) {
	text, ok := value[field].(string)
	if !ok || text == "" {
		return "", fmt.Errorf(`%s field %q must be a non-empty string`, label, field)
	}
	return text, nil
}

func validateClaudeOptionalString(value map[string]any, field, label string) error {
	if item, exists := value[field]; exists {
		if _, ok := item.(string); !ok {
			return fmt.Errorf(`%s field %q must be a string`, label, field)
		}
	}
	return nil
}

func validateClaudeOptionalBoolean(value map[string]any, field, label string) error {
	if item, exists := value[field]; exists {
		if _, ok := item.(bool); !ok {
			return fmt.Errorf(`%s field %q must be a boolean`, label, field)
		}
	}
	return nil
}

func validateClaudeCommonFields(value map[string]any, label string) error {
	for _, field := range []string{"if", "statusMessage"} {
		if err := validateClaudeOptionalString(value, field, label); err != nil {
			return err
		}
	}
	if err := validateClaudeOptionalBoolean(value, "once", label); err != nil {
		return err
	}
	if item, exists := value["timeout"]; exists {
		timeout, ok := item.(float64)
		if !ok || math.IsNaN(timeout) || math.IsInf(timeout, 0) || timeout <= 0 {
			return fmt.Errorf("%s timeout must be a positive finite number", label)
		}
	}
	return nil
}

func validateClaudeStringArray(value any, field, label string, nonEmpty bool) error {
	items, ok := value.([]any)
	if !ok {
		return fmt.Errorf(`%s field %q must be an array of strings`, label, field)
	}
	for _, item := range items {
		text, ok := item.(string)
		if !ok {
			return fmt.Errorf(`%s field %q must be an array of strings`, label, field)
		}
		if nonEmpty && text == "" {
			return fmt.Errorf(
				`%s field %q must contain non-empty strings`,
				label,
				field,
			)
		}
	}
	return nil
}

func validateClaudeHTTPHandler(handler map[string]any, label string) error {
	rawURL, err := requireClaudeString(handler, "url", label)
	if err != nil {
		return err
	}
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return fmt.Errorf(`%s field "url" must be a valid URL`, label)
	}
	if (parsed.Scheme != "http" && parsed.Scheme != "https") || parsed.Hostname() == "" {
		return fmt.Errorf(`%s field "url" must use HTTP or HTTPS`, label)
	}
	if headers, exists := handler["headers"]; exists {
		object, ok := headers.(map[string]any)
		if !ok {
			return fmt.Errorf(`%s field "headers" must map names to strings`, label)
		}
		for _, item := range object {
			if _, ok := item.(string); !ok {
				return fmt.Errorf(`%s field "headers" must map names to strings`, label)
			}
		}
	}
	if allowed, exists := handler["allowedEnvVars"]; exists {
		return validateClaudeStringArray(allowed, "allowedEnvVars", label, true)
	}
	return nil
}

func validateClaudeHandler(
	value any,
	event string,
	groupIndex int,
	handlerIndex int,
) (map[string]any, error) {
	label := fmt.Sprintf(
		"Claude Code settings handler hooks.%s[%d].hooks[%d]",
		event,
		groupIndex,
		handlerIndex,
	)
	handler, ok := value.(map[string]any)
	if !ok || handler == nil {
		return nil, fmt.Errorf("%s must be an object", label)
	}
	handlerType, ok := handler["type"].(string)
	allowed, supported := claudeHandlerKeys[handlerType]
	if !ok || !supported {
		return nil, fmt.Errorf(`%s has unsupported type %q`, label, fmt.Sprint(handler["type"]))
	}
	if err := validateClaudeKnownKeys(handler, allowed, label); err != nil {
		return nil, err
	}
	if err := validateClaudeCommonFields(handler, label); err != nil {
		return nil, err
	}
	switch handlerType {
	case "command":
		if _, err := requireClaudeString(handler, "command", label); err != nil {
			return nil, err
		}
		if args, exists := handler["args"]; exists {
			if err := validateClaudeStringArray(args, "args", label, false); err != nil {
				return nil, err
			}
		}
		for _, field := range []string{"async", "asyncRewake"} {
			if err := validateClaudeOptionalBoolean(handler, field, label); err != nil {
				return nil, err
			}
		}
		for _, field := range []string{"rewakeMessage", "rewakeSummary"} {
			if _, exists := handler[field]; exists {
				if _, err := requireClaudeString(handler, field, label); err != nil {
					return nil, err
				}
			}
		}
		if shell, exists := handler["shell"]; exists &&
			shell != "bash" && shell != "powershell" {
			return nil, fmt.Errorf(
				`%s field "shell" must be "bash" or "powershell"`,
				label,
			)
		}
	case "prompt", "agent":
		if _, err := requireClaudeString(handler, "prompt", label); err != nil {
			return nil, err
		}
		if err := validateClaudeOptionalString(handler, "model", label); err != nil {
			return nil, err
		}
		if handlerType == "prompt" {
			if err := validateClaudeOptionalBoolean(
				handler,
				"continueOnBlock",
				label,
			); err != nil {
				return nil, err
			}
		}
	case "http":
		if err := validateClaudeHTTPHandler(handler, label); err != nil {
			return nil, err
		}
	case "mcp_tool":
		for _, field := range []string{"server", "tool"} {
			if _, err := requireClaudeString(handler, field, label); err != nil {
				return nil, err
			}
		}
		if input, exists := handler["input"]; exists {
			if _, ok := input.(map[string]any); !ok {
				return nil, fmt.Errorf(`%s field "input" must be an object`, label)
			}
		}
	}
	return cloneClaudeObject(handler), nil
}

func validateClaudeGroup(value any, event string, groupIndex int) (claudeGroup, error) {
	label := fmt.Sprintf("Claude Code settings matcher group hooks.%s[%d]", event, groupIndex)
	object, ok := value.(map[string]any)
	if !ok || object == nil {
		return claudeGroup{}, fmt.Errorf("%s must be an object", label)
	}
	if err := validateClaudeKnownKeys(
		object,
		stringSet("hooks", "matcher"),
		label,
	); err != nil {
		return claudeGroup{}, err
	}
	if matcher, exists := object["matcher"]; exists {
		if _, ok := matcher.(string); !ok {
			return claudeGroup{}, fmt.Errorf("%s matcher must be a string", label)
		}
	}
	values, ok := object["hooks"].([]any)
	if !ok {
		return claudeGroup{}, fmt.Errorf("%s must contain a hooks array", label)
	}
	handlers := make([]map[string]any, 0, len(values))
	for handlerIndex, value := range values {
		handler, err := validateClaudeHandler(value, event, groupIndex, handlerIndex)
		if err != nil {
			return claudeGroup{}, err
		}
		handlers = append(handlers, handler)
	}
	return claudeGroup{object: cloneClaudeObject(object), handlers: handlers}, nil
}

func readClaudeHooks(root map[string]any) (claudeHooks, error) {
	value, exists := root["hooks"]
	if !exists {
		return claudeHooks{}, nil
	}
	object, ok := value.(map[string]any)
	if !ok || object == nil {
		return nil, fmt.Errorf(`Claude Code settings field "hooks" must be an object`)
	}
	hooks := make(claudeHooks, len(object))
	for event, groupsValue := range object {
		if _, supported := claudeHookEvents[event]; !supported {
			return nil, fmt.Errorf(
				`Claude Code settings contains unsupported hook event %q`,
				event,
			)
		}
		values, ok := groupsValue.([]any)
		if !ok {
			return nil, fmt.Errorf(
				`Claude Code settings field "hooks.%s" must be an array`,
				event,
			)
		}
		groups := make([]claudeGroup, 0, len(values))
		for index, value := range values {
			group, err := validateClaudeGroup(value, event, index)
			if err != nil {
				return nil, err
			}
			groups = append(groups, group)
		}
		hooks[event] = groups
	}
	return hooks, nil
}

func parseClaudeDocument(filePath string, raw []byte) (*claudeDocument, error) {
	label := fmt.Sprintf("Claude Code user settings at %s", filePath)
	root, err := decodeStrictJSONObject(raw, label)
	if err != nil {
		return nil, err
	}
	disabled := false
	if value, exists := root["disableAllHooks"]; exists {
		var ok bool
		disabled, ok = value.(bool)
		if !ok {
			return nil, fmt.Errorf(
				`Claude Code settings field "disableAllHooks" must be a boolean`,
			)
		}
	}
	hooks, err := readClaudeHooks(root)
	if err != nil {
		return nil, err
	}
	return &claudeDocument{
		exists: true, filePath: filePath, root: root, hooks: hooks,
		hooksDisabled: disabled, raw: append([]byte(nil), raw...),
	}, nil
}

func createClaudeDocument(filePath string) *claudeDocument {
	return &claudeDocument{
		filePath: filePath,
		root:     map[string]any{},
		hooks:    claudeHooks{},
	}
}

func renderClaudeHooks(hooks claudeHooks) map[string]any {
	result := make(map[string]any, len(hooks))
	for event, groups := range hooks {
		values := make([]any, 0, len(groups))
		for _, group := range groups {
			object := cloneClaudeObject(group.object)
			handlers := make([]any, 0, len(group.handlers))
			for _, handler := range group.handlers {
				handlers = append(handlers, handler)
			}
			object["hooks"] = handlers
			values = append(values, object)
		}
		result[event] = values
	}
	return result
}

func renderClaudeDocument(
	document *claudeDocument,
	hooks claudeHooks,
) (*claudeRenderedDocument, error) {
	if !document.exists && len(hooks) == 0 {
		return &claudeRenderedDocument{document: document}, nil
	}
	if reflect.DeepEqual(hooks, document.hooks) {
		return &claudeRenderedDocument{document: document}, nil
	}
	if len(hooks) == 0 && entirelyManagedClaudeDocument(document) {
		return &claudeRenderedDocument{
			document: document, changed: true, remove: true,
		}, nil
	}
	root := cloneClaudeObject(document.root)
	if len(hooks) == 0 {
		delete(root, "hooks")
	} else {
		root["hooks"] = renderClaudeHooks(hooks)
	}
	next, err := json.MarshalIndent(root, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("encode Claude Code user settings: %w", err)
	}
	next = append(next, '\n')
	if _, err := parseClaudeDocument(document.filePath, next); err != nil {
		return nil, fmt.Errorf("validate rendered Claude Code user settings: %w", err)
	}
	return &claudeRenderedDocument{
		document: document,
		changed:  !document.exists || !bytes.Equal(next, document.raw),
		next:     next,
	}, nil
}
